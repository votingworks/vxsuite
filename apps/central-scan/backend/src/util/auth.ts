import {
  DippedSmartCardAuthApi,
  DippedSmartCardAuthMachineState,
} from '@votingworks/auth';
import { DEFAULT_SYSTEM_SETTINGS } from '@votingworks/types';
import { LoggingUserRole } from '@votingworks/logging';
import { Workspace } from './workspace';

export function constructAuthMachineState(
  workspace: Workspace
): DippedSmartCardAuthMachineState {
  const electionKey = workspace.store.getElectionKey();
  const jurisdiction = workspace.store.getJurisdiction();
  const systemSettings =
    workspace.store.getSystemSettings() ?? DEFAULT_SYSTEM_SETTINGS;
  return {
    ...systemSettings.auth,
    electionKey,
    jurisdiction,
  };
}

export async function getUserRole(
  auth: DippedSmartCardAuthApi,
  workspace: Workspace
): Promise<LoggingUserRole> {
  const authStatus = await auth.getAuthStatus(
    constructAuthMachineState(workspace)
  );
  return authStatus.status === 'logged_in' ? authStatus.user.role : 'unknown';
}
