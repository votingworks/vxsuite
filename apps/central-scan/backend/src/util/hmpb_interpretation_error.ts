export class HmpbInterpretationError extends Error {}
