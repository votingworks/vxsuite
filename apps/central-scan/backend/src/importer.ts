import { Scan } from '@votingworks/api';
import { Result, assert, ok, sleep } from '@votingworks/basics';
import {
  ElectionDefinition,
  MarkThresholds,
  PageInterpretation,
  PageInterpretationWithFiles,
  SheetOf,
} from '@votingworks/types';
import { ALL_PRECINCTS_SELECTION } from '@votingworks/utils';
import makeDebug from 'debug';
import * as fsExtra from 'fs-extra';
import { join } from 'path';
import { v4 as uuid } from 'uuid';
import { interpretSheetAndSaveImages } from '@votingworks/ballot-interpreter';
import { BatchControl, BatchScanner } from './fujitsu_scanner';
import { Castability, checkSheetCastability } from './util/castability';
import { Workspace } from './util/workspace';
import {
  describeValidationError,
  validateSheetInterpretation,
} from './validation';

const debug = makeDebug('scan:importer');

export interface Options {
  workspace: Workspace;
  scanner: BatchScanner;
}

/**
 * Imports ballot images from a `Scanner` and stores them in a `Store`.
 */
export class Importer {
  private readonly workspace: Workspace;
  private readonly scanner: BatchScanner;
  private sheetGenerator?: BatchControl;
  private batchId?: string;

  constructor({ workspace, scanner }: Options) {
    this.workspace = workspace;
    this.scanner = scanner;
  }

  /**
   * Sets the election information used to encode and decode ballots.
   */
  configure(
    electionDefinition: ElectionDefinition,
    jurisdiction: string
  ): void {
    this.workspace.store.setElectionAndJurisdiction({
      electionData: electionDefinition.electionData,
      jurisdiction,
    });
    // Central scanner only uses all precinct mode, set on every configure
    this.workspace.store.setPrecinctSelection(ALL_PRECINCTS_SELECTION);
  }

  setTestMode(testMode: boolean): void {
    debug('setting test mode to %s', testMode);
    this.doZero();
    this.workspace.store.setTestMode(testMode);
  }

  setMarkThresholdOverrides(markThresholds?: MarkThresholds): void {
    debug('setting mark thresholds overrides to %s', markThresholds);
    this.workspace.store.setMarkThresholdOverrides(markThresholds);
  }

  private async sheetAdded(
    paths: SheetOf<string>,
    batchId: string
  ): Promise<string> {
    const start = Date.now();
    try {
      debug('sheetAdded %o batchId=%s STARTING', paths, batchId);
      return await this.importSheet(batchId, paths[0], paths[1]);
    } finally {
      const end = Date.now();
      debug(
        'sheetAdded %o batchId=%s FINISHED in %dms',
        paths,
        batchId,
        Math.round(end - start)
      );
    }
  }

  async importSheet(
    batchId: string,
    frontInputImagePath: string,
    backInputImagePath: string
  ): Promise<string> {
    let sheetId = uuid();
    const interpretResult = await this.interpretSheet(sheetId, [
      frontInputImagePath,
      backInputImagePath,
    ]);

    if (interpretResult.isErr()) {
      throw interpretResult.err();
    }

    const [{ imagePath: frontImagePath }, { imagePath: backImagePath }] =
      interpretResult.ok();
    let [
      { interpretation: frontInterpretation },
      { interpretation: backInterpretation },
    ] = interpretResult.ok();

    debug(
      'interpreted %s (%s): %O',
      frontImagePath,
      frontInterpretation.type,
      frontInterpretation
    );
    debug(
      'interpreted %s (%s): %O',
      backImagePath,
      backInterpretation.type,
      backInterpretation
    );

    const validationResult = validateSheetInterpretation([
      frontInterpretation,
      backInterpretation,
    ]);
    if (validationResult.isErr()) {
      const error = validationResult.err();
      const errDescription = describeValidationError(error);
      debug(
        'rejecting sheet because it would not produce a valid CVR: error=%s: %o',
        errDescription,
        error
      );
      // replaces interpretation with something that cannot be accepted
      frontInterpretation = {
        type: 'UnreadablePage',
        reason: `invalid CVR: ${errDescription}`,
      };
      backInterpretation = {
        type: 'UnreadablePage',
        reason: `invalid CVR: ${errDescription}`,
      };
    }

    sheetId = await this.addSheet(
      batchId,
      frontImagePath,
      frontInterpretation,
      backImagePath,
      backInterpretation
    );

    return sheetId;
  }

  private async interpretSheet(
    sheetId: string,
    [frontImagePath, backImagePath]: SheetOf<string>
  ): Promise<Result<SheetOf<PageInterpretationWithFiles>, Error>> {
    const electionDefinition = this.getElectionDefinition();

    return ok(
      await interpretSheetAndSaveImages(
        {
          electionDefinition,
          precinctSelection: ALL_PRECINCTS_SELECTION,
          testMode: this.workspace.store.getTestMode(),
          adjudicationReasons:
            electionDefinition.election.centralScanAdjudicationReasons,
          markThresholds: this.workspace.store.getMarkThresholdOverrides(),
        },
        [frontImagePath, backImagePath],
        sheetId,
        this.workspace.ballotImagesPath
      )
    );
  }

  /**
   * Add a sheet to the internal store.
   */
  private async addSheet(
    batchId: string,
    frontImagePath: string,
    frontInterpretation: PageInterpretation,
    backImagePath: string,
    backInterpretation: PageInterpretation
  ): Promise<string> {
    if ('metadata' in frontInterpretation && 'metadata' in backInterpretation) {
      if (
        'pageNumber' in frontInterpretation.metadata &&
        'pageNumber' in backInterpretation.metadata
      ) {
        if (
          frontInterpretation.metadata.pageNumber >
          backInterpretation.metadata.pageNumber
        ) {
          return this.addSheet(
            batchId,
            backImagePath,
            backInterpretation,
            frontImagePath,
            frontInterpretation
          );
        }
      }
    }

    const ballotId = this.workspace.store.addSheet(uuid(), batchId, [
      {
        imagePath: frontImagePath,
        interpretation: frontInterpretation,
      },
      {
        imagePath: backImagePath,
        interpretation: backInterpretation,
      },
    ]);

    return ballotId;
  }

  private async finishBatch(error?: string): Promise<void> {
    if (this.batchId) {
      this.workspace.store.finishBatch({ batchId: this.batchId, error });
      this.batchId = undefined;
    }

    if (this.sheetGenerator) {
      await this.sheetGenerator.endBatch();
      this.sheetGenerator = undefined;
    }
  }

  /**
   * Scan a single sheet and see how it looks
   */
  private async scanOneSheet(): Promise<void> {
    assert(
      typeof this.sheetGenerator !== 'undefined' &&
        typeof this.batchId !== 'undefined'
    );

    const sheet = await this.sheetGenerator.scanSheet();

    if (!sheet) {
      debug('closing batch %s', this.batchId);
      await this.finishBatch();
    } else {
      debug('got a ballot card: %o', sheet);
      const sheetId = await this.sheetAdded(sheet, this.batchId);
      debug('got a ballot card: %o, %s', sheet, sheetId);

      const adjudicationStatus = this.workspace.store.adjudicationStatus();
      if (adjudicationStatus.remaining === 0) {
        if (!(await this.sheetGenerator.acceptSheet())) {
          debug('failed to accept interpreted sheet: %s', sheetId);
        }
        await this.continueImport({ forceAccept: false });
      } else {
        const castability = this.getNextAdjudicationCastability();
        if (castability) {
          if (castability === Castability.Uncastable) {
            await this.sheetGenerator.rejectSheet();
          } else {
            await this.sheetGenerator.reviewSheet();
          }
        }
      }
    }
  }

  getNextAdjudicationCastability(): Castability | undefined {
    const sheet = this.workspace.store.getNextAdjudicationSheet();
    if (sheet) {
      return checkSheetCastability([
        sheet.front.interpretation,
        sheet.back.interpretation,
      ]);
    }
  }

  /**
   * Create a new batch and begin the scanning process
   */
  async startImport(): Promise<string> {
    this.getElectionDefinition(); // ensure election definition is loaded

    if (this.sheetGenerator) {
      throw new Error('scanning already in progress');
    }

    this.batchId = this.workspace.store.addBatch();
    const batchScanDirectory = join(
      this.workspace.ballotImagesPath,
      `batch-${this.batchId}`
    );
    await fsExtra.ensureDir(batchScanDirectory);
    debug(
      'scanning starting for batch %s into %s',
      this.batchId,
      batchScanDirectory
    );
    const ballotPaperSize =
      this.workspace.store.getBallotPaperSizeForElection();
    this.sheetGenerator = this.scanner.scanSheets({
      directory: batchScanDirectory,
      pageSize: ballotPaperSize,
    });

    await this.continueImport({ forceAccept: false });

    return this.batchId;
  }

  /**
   * Continue the existing scanning process
   */
  async continueImport(options: { forceAccept: boolean }): Promise<void> {
    const sheet = this.workspace.store.getNextAdjudicationSheet();

    if (sheet) {
      if (options.forceAccept) {
        await this.sheetGenerator?.acceptSheet();
        this.workspace.store.adjudicateSheet(sheet.id);
      } else {
        await this.sheetGenerator?.rejectSheet();
        this.workspace.store.deleteSheet(sheet.id);
      }
    }

    if (this.sheetGenerator && this.batchId) {
      this.scanOneSheet().catch((error) => {
        debug('processing sheet failed with error: %s', error.stack);
        void this.finishBatch(error.toString());
      });
    } else {
      throw new Error('no scanning job in progress');
    }
  }

  /**
   * this is really for testing
   */
  async waitForEndOfBatchOrScanningPause(): Promise<void> {
    if (!this.batchId) {
      return;
    }

    const adjudicationStatus = this.workspace.store.adjudicationStatus();
    if (adjudicationStatus.remaining > 0) {
      return;
    }

    await sleep(200);
    return this.waitForEndOfBatchOrScanningPause();
  }

  /**
   * Reset all the data, both in the store and the ballot images.
   */
  doZero(): void {
    this.workspace.resetElectionSession();
  }

  /**
   * Get the imported batches and current election info, if any.
   */
  getStatus(): Scan.ScanStatus {
    const electionDefinition = this.getElectionDefinition();
    const canUnconfigure = this.workspace.store.getCanUnconfigure();
    const batches = this.workspace.store.batchStatus();
    const adjudication = this.workspace.store.adjudicationStatus();

    return {
      electionHash: electionDefinition?.electionHash,
      batches,
      canUnconfigure,
      adjudication,
    };
  }

  /**
   * Resets all data like `doZero`, removes election info, and stops importing.
   */
  unconfigure(): void {
    this.doZero();
    this.workspace.store.reset(); // destroy all data
  }

  private getElectionDefinition(): ElectionDefinition {
    const electionDefinition = this.workspace.store.getElectionDefinition();

    if (!electionDefinition) {
      throw new Error('no election configuration');
    }

    return electionDefinition;
  }
}
