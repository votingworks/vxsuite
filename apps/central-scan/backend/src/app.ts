import { Scan } from '@votingworks/api';
import {
  DippedSmartCardAuthApi,
  generateSignedHashValidationQrCodeValue,
} from '@votingworks/auth';
import { Result, assert, ok } from '@votingworks/basics';
import {
  createSystemCallApi,
  DiskSpaceSummary,
  readSignedElectionPackageFromUsb,
  exportCastVoteRecordsToUsbDrive,
  ElectionRecord,
} from '@votingworks/backend';
import {
  ElectionPackageConfigurationError,
  BallotPageLayout,
  DEFAULT_SYSTEM_SETTINGS,
  ElectionDefinition,
  SystemSettings,
  ExportCastVoteRecordsToUsbDriveError,
  DiagnosticRecord,
} from '@votingworks/types';
import { isElectionManagerAuth } from '@votingworks/utils';
import express, { Application } from 'express';
import * as grout from '@votingworks/grout';
import { LogEventId, Logger } from '@votingworks/logging';
import { useDevDockRouter } from '@votingworks/dev-dock-backend';
import { UsbDrive, UsbDriveStatus } from '@votingworks/usb-drive';
import { Importer } from './importer';
import { Workspace } from './util/workspace';
import { MachineConfig, ScanStatus } from './types';
import { getMachineConfig } from './machine_config';
import { constructAuthMachineState } from './util/auth';
import {
  logBatchStartFailure,
  logBatchStartSuccess,
  logScanBatchContinueFailure,
  logScanBatchContinueSuccess,
} from './util/logging';
import { saveReadinessReport } from './readiness_report';
import { performScanDiagnostic, ScanDiagnosticOutcome } from './diagnostic';
import { BatchScanner } from './fujitsu_scanner';

type NoParams = never;

export interface AppOptions {
  auth: DippedSmartCardAuthApi;
  allowedExportPatterns?: string[];
  scanner: BatchScanner;
  importer: Importer;
  workspace: Workspace;
  logger: Logger;
  usbDrive: UsbDrive;
}

function buildApi({
  auth,
  workspace,
  logger,
  usbDrive,
  scanner,
  importer,
}: Exclude<AppOptions, 'allowedExportPatterns'>) {
  const { store } = workspace;

  return grout.createApi({
    getAuthStatus() {
      return auth.getAuthStatus(constructAuthMachineState(workspace));
    },

    checkPin(input: { pin: string }) {
      return auth.checkPin(constructAuthMachineState(workspace), input);
    },

    logOut() {
      return auth.logOut(constructAuthMachineState(workspace));
    },

    getUsbDriveStatus(): Promise<UsbDriveStatus> {
      return usbDrive.status();
    },

    async ejectUsbDrive(): Promise<void> {
      return usbDrive.eject();
    },

    getMachineConfig(): MachineConfig {
      return getMachineConfig();
    },

    getTestMode() {
      return store.getTestMode();
    },

    async setTestMode(input: { testMode: boolean }) {
      const { testMode } = input;
      await logger.logAsCurrentRole(LogEventId.TogglingTestMode, {
        message: `Toggling to ${testMode ? 'Test' : 'Official'} Ballot Mode...`,
      });
      await importer.setTestMode(testMode);
      await logger.logAsCurrentRole(LogEventId.ToggledTestMode, {
        disposition: 'success',
        message: `Successfully toggled to ${
          testMode ? 'Test' : 'Official'
        } Ballot Mode`,
      });
    },

    updateSessionExpiry(input: { sessionExpiresAt: Date }) {
      return auth.updateSessionExpiry(
        constructAuthMachineState(workspace),
        input
      );
    },

    async deleteBatch({ batchId }: { batchId: string }) {
      const numberOfBallotsInBatch = workspace.store
        .getBatches()
        .find((batch) => batch.id === batchId)?.count;

      await logger.logAsCurrentRole(LogEventId.DeleteScanBatchInit, {
        message: `User deleting batch id ${batchId}...`,
        numberOfBallotsInBatch,
        batchId,
      });

      try {
        workspace.store.deleteBatch(batchId);
        await logger.logAsCurrentRole(LogEventId.DeleteScanBatchComplete, {
          disposition: 'success',
          message: `User successfully deleted batch id: ${batchId} containing ${numberOfBallotsInBatch} ballots.`,
          numberOfBallotsInBatch,
          batchId,
        });
      } catch (error) {
        assert(error instanceof Error);
        await logger.logAsCurrentRole(LogEventId.DeleteScanBatchComplete, {
          disposition: 'failure',
          message: `Error deleting batch id: ${batchId}.`,
          error: error.message,
          result: 'Batch not deleted.',
        });
        throw error;
      }
    },

    async configureFromElectionPackageOnUsbDrive(): Promise<
      Result<ElectionDefinition, ElectionPackageConfigurationError>
    > {
      const authStatus = await auth.getAuthStatus(
        constructAuthMachineState(workspace)
      );

      const electionPackageResult = await readSignedElectionPackageFromUsb(
        authStatus,
        usbDrive,
        logger
      );
      if (electionPackageResult.isErr()) {
        await logger.logAsCurrentRole(LogEventId.ElectionConfigured, {
          message: `Error configuring machine.`,
          disposition: 'failure',
          errorDetails: JSON.stringify(electionPackageResult.err()),
        });
        return electionPackageResult;
      }
      assert(isElectionManagerAuth(authStatus));
      const { electionPackage, electionPackageHash } =
        electionPackageResult.ok();
      const { electionDefinition, systemSettings } = electionPackage;
      assert(systemSettings);

      importer.configure(
        electionDefinition,
        authStatus.user.jurisdiction,
        electionPackageHash
      );
      store.setSystemSettings(systemSettings);

      await logger.logAsCurrentRole(LogEventId.ElectionConfigured, {
        message: `Machine configured for election with hash: ${electionDefinition.ballotHash}`,
        disposition: 'success',
        ballotHash: electionDefinition.ballotHash,
      });

      return ok(electionDefinition);
    },

    getSystemSettings(): SystemSettings {
      return workspace.store.getSystemSettings() ?? DEFAULT_SYSTEM_SETTINGS;
    },

    getElectionRecord(): ElectionRecord | null {
      return store.getElectionRecord() || null;
    },

    getStatus(): ScanStatus {
      return importer.getStatus();
    },

    async scanBatch(): Promise<void> {
      try {
        const batchId = await importer.startImport();
        await logBatchStartSuccess(logger, batchId);
      } catch (error) {
        assert(error instanceof Error);
        await logBatchStartFailure(logger, error);
      }
    },

    async continueScanning(input: { forceAccept: boolean }): Promise<void> {
      try {
        const { forceAccept } = input;
        importer.continueImport(input);
        await logScanBatchContinueSuccess(logger, forceAccept);
      } catch (error) {
        assert(error instanceof Error);
        await logScanBatchContinueFailure(logger, error);
      }
    },

    async unconfigure(
      input: {
        ignoreBackupRequirement?: boolean;
      } = {}
    ): Promise<void> {
      // frontend should only allow this call if the machine can be unconfigured
      assert(store.getCanUnconfigure() || input.ignoreBackupRequirement);

      await importer.unconfigure();
      await logger.logAsCurrentRole(LogEventId.ElectionUnconfigured, {
        disposition: 'success',
        message:
          'User successfully unconfigured the machine to remove the current election and all current ballot data.',
      });
    },

    async clearBallotData(): Promise<void> {
      // frontend should only allow this call if the machine can be unconfigured
      assert(store.getCanUnconfigure());

      await importer.doZero();
    },

    async exportCastVoteRecordsToUsbDrive(input: {
      isMinimalExport?: boolean;
    }): Promise<Result<void, ExportCastVoteRecordsToUsbDriveError>> {
      const logItem = input.isMinimalExport ? 'cast vote records' : 'backup';
      await logger.logAsCurrentRole(LogEventId.ExportCastVoteRecordsInit, {
        message: `Exporting ${logItem}...`,
      });
      const exportResult = await exportCastVoteRecordsToUsbDrive(
        store,
        usbDrive,
        input.isMinimalExport
          ? store.forEachAcceptedSheet()
          : store.forEachSheet(),
        { scannerType: 'central', isMinimalExport: input.isMinimalExport }
      );
      if (!input.isMinimalExport) {
        store.setScannerBackedUp();
      }
      if (exportResult.isErr()) {
        await logger.logAsCurrentRole(
          LogEventId.ExportCastVoteRecordsComplete,
          {
            disposition: 'failure',
            message: `Error exporting ${logItem}.`,
            errorDetails: JSON.stringify(exportResult.err()),
          }
        );
      } else {
        await logger.logAsCurrentRole(
          LogEventId.ExportCastVoteRecordsComplete,
          {
            disposition: 'success',
            message: `Successfully exported ${logItem}.`,
          }
        );
      }
      return exportResult;
    },

    saveReadinessReport() {
      return saveReadinessReport({
        workspace,
        isScannerAttached: importer.getStatus().isScannerAttached,
        usbDrive,
        logger,
      });
    },

    async performScanDiagnostic(): Promise<ScanDiagnosticOutcome> {
      return await performScanDiagnostic(scanner, store, logger);
    },

    getMostRecentScannerDiagnostic(): DiagnosticRecord | null {
      return store.getMostRecentDiagnosticRecord('blank-sheet-scan') ?? null;
    },

    async getApplicationDiskSpaceSummary(): Promise<DiskSpaceSummary> {
      return workspace.getDiskSpaceSummary();
    },

    /* istanbul ignore next */
    async generateSignedHashValidationQrCodeValue() {
      const { codeVersion, machineId } = getMachineConfig();
      const electionRecord = store.getElectionRecord();
      await logger.logAsCurrentRole(LogEventId.SignedHashValidationInit);
      const qrCodeValue = await generateSignedHashValidationQrCodeValue({
        electionRecord,
        machineId,
        softwareVersion: codeVersion,
      });
      await logger.logAsCurrentRole(LogEventId.SignedHashValidationComplete, {
        disposition: 'success',
      });
      return qrCodeValue;
    },

    ...createSystemCallApi({
      usbDrive,
      logger,
      machineId: getMachineConfig().machineId,
      codeVersion: getMachineConfig().codeVersion,
    }),
  });
}

/**
 * A type to be used by the frontend to create a Grout API client
 */
export type Api = ReturnType<typeof buildApi>;

/**
 * Builds an express application, using `store` and `importer` to do the heavy
 * lifting.
 */
export function buildCentralScannerApp({
  auth,
  scanner,
  importer,
  workspace,
  logger,
  usbDrive,
}: AppOptions): Application {
  const { store } = workspace;

  const app: Application = express();
  const api = buildApi({
    auth,
    workspace,
    logger,
    usbDrive,
    scanner,
    importer,
  });
  app.use('/api', grout.buildRouter(api, express));
  useDevDockRouter(app, express, 'central-scan');

  const deprecatedApiRouter = express.Router();
  deprecatedApiRouter.use(express.raw());
  deprecatedApiRouter.use(
    express.json({ limit: '5mb', type: 'application/json' })
  );
  deprecatedApiRouter.use(express.urlencoded({ extended: false }));

  deprecatedApiRouter.get(
    '/central-scanner/scan/hmpb/ballot/:sheetId/:side/image',
    (request, response) => {
      const { sheetId, side } = request.params;

      if (
        typeof sheetId !== 'string' ||
        (side !== 'front' && side !== 'back')
      ) {
        response.status(404);
        return;
      }
      const imagePath = store.getBallotImagePath(sheetId, side);

      if (imagePath) {
        response.sendFile(imagePath);
      } else {
        response.status(404).end();
      }
    }
  );

  deprecatedApiRouter.get<NoParams, Scan.GetNextReviewSheetResponse>(
    '/central-scanner/scan/hmpb/review/next-sheet',
    (_request, response) => {
      const sheet = store.getNextAdjudicationSheet();

      if (sheet) {
        let frontLayout: BallotPageLayout | undefined;
        let backLayout: BallotPageLayout | undefined;
        let frontDefinition:
          | Scan.GetNextReviewSheetResponse['definitions']['front']
          | undefined;
        let backDefinition:
          | Scan.GetNextReviewSheetResponse['definitions']['back']
          | undefined;

        if (sheet.front.interpretation.type === 'InterpretedHmpbPage') {
          const front = sheet.front.interpretation;
          frontLayout = front.layout;
          const contestIds = Object.keys(front.votes);
          frontDefinition = { contestIds };
        }

        if (sheet.back.interpretation.type === 'InterpretedHmpbPage') {
          const back = sheet.back.interpretation;
          const contestIds = Object.keys(back.votes);

          backLayout = back.layout;
          backDefinition = { contestIds };
        }

        response.json({
          interpreted: sheet,
          layouts: {
            front: frontLayout,
            back: backLayout,
          },
          definitions: {
            front: frontDefinition,
            back: backDefinition,
          },
        });
      } else {
        response.status(404).end();
      }
    }
  );

  app.use(deprecatedApiRouter);

  return app;
}
