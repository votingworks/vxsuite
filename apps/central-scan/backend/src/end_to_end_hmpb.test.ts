import getPort from 'get-port';
import {
  getCastVoteRecordExportDirectoryPaths,
  isTestReport,
  mockElectionPackageFileTree,
  readCastVoteRecordExport,
} from '@votingworks/backend';
import {
  asElectionDefinition,
  electionGridLayoutNewHampshireTestBallotFixtures,
} from '@votingworks/fixtures';
import {
  CVR,
  DEFAULT_SYSTEM_SETTINGS,
  TEST_JURISDICTION,
} from '@votingworks/types';
import * as grout from '@votingworks/grout';
import {
  BooleanEnvironmentVariableName,
  convertCastVoteRecordVotesToTabulationVotes,
  getFeatureFlagMock,
} from '@votingworks/utils';
import { EventEmitter } from 'events';
import { Application } from 'express';
import * as fs from 'fs-extra';
import { join } from 'path';
import request from 'supertest';
import { dirSync } from 'tmp';
import { buildMockDippedSmartCardAuth } from '@votingworks/auth';
import { fakeLogger, Logger } from '@votingworks/logging';
import { Server } from 'http';
import { fakeSessionExpiresAt } from '@votingworks/test-utils';
import { ok } from '@votingworks/basics';
import { MockUsbDrive, createMockUsbDrive } from '@votingworks/usb-drive';
import { makeMockScanner, MockScanner } from '../test/util/mocks';
import { Importer } from './importer';
import { createWorkspace, Workspace } from './util/workspace';
import { Api, buildCentralScannerApp } from './app';
import { start } from './server';

// mock SKIP_SCAN_ELECTION_HASH_CHECK to allow us to use old ballot image fixtures
const featureFlagMock = getFeatureFlagMock();
jest.mock('@votingworks/utils', () => {
  return {
    ...jest.requireActual('@votingworks/utils'),
    isFeatureFlagEnabled: (flag: BooleanEnvironmentVariableName) =>
      featureFlagMock.isEnabled(flag),
  };
});

jest.mock('./exec', () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/require-await
  default: async (): Promise<{ stdout: string; stderr: string }> => ({
    stdout: '',
    stderr: '',
  }),
  streamExecFile: (): unknown => {
    const child = new EventEmitter();

    Object.defineProperties(child, {
      stdout: { value: new EventEmitter() },
      stderr: { value: new EventEmitter() },
    });

    process.nextTick(() => child.emit('exit', 0));

    return child;
  },
}));

let auth: ReturnType<typeof buildMockDippedSmartCardAuth>;
let workspace: Workspace;
let scanner: MockScanner;
let mockUsbDrive: MockUsbDrive;
let importer: Importer;
let app: Application;
let logger: Logger;
let apiClient: grout.Client<Api>;
let server: Server;

beforeEach(async () => {
  const port = await getPort();
  auth = buildMockDippedSmartCardAuth();
  workspace = createWorkspace(dirSync().name);
  scanner = makeMockScanner();
  importer = new Importer({ workspace, scanner });
  mockUsbDrive = createMockUsbDrive();
  logger = fakeLogger();
  app = buildCentralScannerApp({
    auth,
    usbDrive: mockUsbDrive.usbDrive,
    allowedExportPatterns: ['/tmp/**'],
    importer,
    workspace,
    logger,
  });
  const baseUrl = `http://localhost:${port}/api`;
  apiClient = grout.createClient({
    baseUrl,
  });

  server = await start({
    app,
    logger,
    workspace,
    port,
  });
});

afterEach(async () => {
  importer.unconfigure();
  await fs.remove(workspace.path);
  featureFlagMock.resetFeatureFlags();
  server.close();
});

const jurisdiction = TEST_JURISDICTION;

test('going through the whole process works', async () => {
  const { election, electionDefinition } =
    electionGridLayoutNewHampshireTestBallotFixtures;

  auth.getAuthStatus.mockResolvedValue({
    status: 'logged_in',
    user: {
      role: 'election_manager',
      jurisdiction,
      electionHash: electionDefinition.electionHash,
    },
    sessionExpiresAt: fakeSessionExpiresAt(),
  });

  importer.configure(asElectionDefinition(election), jurisdiction);
  workspace.store.setSystemSettings(DEFAULT_SYSTEM_SETTINGS);

  // sample ballot election hash does not match election hash for this test
  featureFlagMock.enableFeatureFlag(
    BooleanEnvironmentVariableName.SKIP_SCAN_ELECTION_HASH_CHECK
  );

  featureFlagMock.enableFeatureFlag(
    BooleanEnvironmentVariableName.SKIP_ELECTION_PACKAGE_AUTHENTICATION
  );

  mockUsbDrive.insertUsbDrive(
    await mockElectionPackageFileTree(
      electionGridLayoutNewHampshireTestBallotFixtures.electionJson.toElectionPackage(
        {
          ...DEFAULT_SYSTEM_SETTINGS,
          markThresholds: {
            definite: 0.08,
            marginal: 0.05,
          },
        }
      )
    )
  );
  const configureResult =
    await apiClient.configureFromElectionPackageOnUsbDrive();
  expect(configureResult.err()).toBeUndefined();
  expect(configureResult.isOk()).toEqual(true);
  expect(configureResult.ok()).toEqual(electionDefinition);

  // need to turn off test mode after election is loaded
  await apiClient.setTestMode({ testMode: false });

  mockUsbDrive.removeUsbDrive();
  // Confirm USB drive removed. Must poll the mock USB drive in order to
  // advance the state of the underlying mock function for rest of test.
  expect((await mockUsbDrive.usbDrive.status()).status).toEqual('no_drive');

  {
    // define the next scanner session
    const nextSession = scanner.withNextScannerSession();

    // scan some sample ballots
    nextSession.sheet([
      electionGridLayoutNewHampshireTestBallotFixtures.scanMarkedFront.asFilePath(),
      electionGridLayoutNewHampshireTestBallotFixtures.scanMarkedBack.asFilePath(),
    ]);

    nextSession.end();

    await request(app)
      .post('/central-scanner/scan/scanBatch')
      .expect(200)
      .then((response) => {
        expect(response.body).toEqual({
          status: 'ok',
          batchId: expect.any(String),
        });
      });

    await importer.waitForEndOfBatchOrScanningPause();

    // check the latest batch has the expected counts
    const status = await request(app)
      .get('/central-scanner/scan/status')
      .set('Accept', 'application/json')
      .expect(200);
    expect(JSON.parse(status.text).batches.length).toEqual(1);
    expect(JSON.parse(status.text).batches[0].count).toEqual(1);
  }

  {
    const mockUsbMountPoint = join(workspace.path, 'mock-usb');
    await fs.mkdir(mockUsbMountPoint, { recursive: true });
    mockUsbDrive.insertUsbDrive({});

    expect(
      await apiClient.exportCastVoteRecordsToUsbDrive({
        isMinimalExport: true,
      })
    ).toEqual(ok());

    const cvrReportDirectoryPath = (
      await getCastVoteRecordExportDirectoryPaths(mockUsbDrive.usbDrive)
    )[0];
    expect(cvrReportDirectoryPath).toContain('machine_000__');

    const { castVoteRecordExportMetadata, castVoteRecordIterator } = (
      await readCastVoteRecordExport(cvrReportDirectoryPath)
    ).unsafeUnwrap();
    const cvrs: CVR.CVR[] = (await castVoteRecordIterator.toArray()).map(
      (castVoteRecordResult) =>
        castVoteRecordResult.unsafeUnwrap().castVoteRecord
    );
    expect(cvrs).toHaveLength(1);
    const [cvr] = cvrs;
    expect(
      isTestReport(castVoteRecordExportMetadata.castVoteRecordReportMetadata)
    ).toBeFalsy();
    expect(cvr.BallotStyleId).toEqual('ballot-style-c43745c4');
    expect(cvr.BallotStyleUnitId).toEqual('town-id-00701-precinct');
    expect(cvr.CreatingDeviceId).toEqual('000');
    expect(cvr.BallotSheetId).toEqual('1');
    expect(cvr.vxBallotType).toEqual(CVR.vxBallotType.Precinct);
    expect(
      convertCastVoteRecordVotesToTabulationVotes(cvr.CVRSnapshot[0])
    ).toMatchSnapshot();
  }
}, 25_000);
