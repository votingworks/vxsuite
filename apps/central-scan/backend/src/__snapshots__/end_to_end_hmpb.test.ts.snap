// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`going through the whole process works 1`] = `
{
  "County-Attorney-0b0191d7": [
    "Mary-Woolson-dc0b854a",
  ],
  "County-Commissioner-8ae8f486": [
    "write-in-0",
  ],
  "County-Treasurer-f56fa715": [
    "write-in-0",
  ],
  "Executive-Councilor-17389b9b": [
    "write-in-0",
  ],
  "Governor-30bbdb5e": [
    "Josiah-Bartlett-1bb99985",
  ],
  "Register-of-Deeds-a79861e8": [
    "John-Mann-b56bbdd3",
  ],
  "Register-of-Probate-1941ae43": [
    "Claire-Cutts-07a436e7",
  ],
  "Representative-in-Congress-8aabf59f": [
    "Richard-Coote-b9095636",
  ],
  "Shall-there-be-a-convention-to-amend-or-revise-the-constitution--15e8b5bc": [
    "Shall-there-be-a-convention-to-amend-or-revise-the-constitution--15e8b5bc-option-yes",
  ],
  "Sheriff-0f76c952": [
    "Edward-Randolph-bf4c848a",
  ],
  "State-Representative-Hillsborough-District-37-ade6e70c": [
    "Charles-H-Hersey-096286a4",
  ],
  "State-Representatives-Hillsborough-District-34-9bcd590d": [
    "Samuel-Bell-17973275",
    "Samuel-Livermore-f927fef1",
    "Jacob-Freese-b5146505",
  ],
  "State-Senator-02b21350": [
    "James-Poole-db5ef4bd",
  ],
  "United-States-Senator-839dd0b5": [
    "William-Preston-3778fcd5",
  ],
}
`;
